### ISSUE TITLE
AggregateException in ServiceHub.RoslynCodeAnalysisService32.exe

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn#45749]
I repeatedly see a gold bar while typing in IDE (Roslyn.sln). The call stack in the goldbar is not useful, but servicehub logs shows the below stack:

```
07/07/2020 09:53:31 Pacific Standard Time: Error : 10 :System.AggregateException: One or more errors occurred. ---> System.IO.IOException: Pipe is broken.
   at System.IO.Pipes.PipeStream.CheckWriteOperations()
   at System.IO.Pipes.PipeStream.Write(Byte[] buffer, Int32 offset, Int32 count)
   at System.IO.BufferedStream.FlushWrite()
   at System.IO.BufferedStream.Flush()
   at System.IO.BufferedStream.Dispose(Boolean disposing)
   at System.IO.Stream.Close()
   at System.IO.Stream.Dispose()
   at Microsoft.CodeAnalysis.Remote.RemoteEndPoint.<WriteDataToNamedPipeAsync>d__23`1.MoveNext() in C:\roslyn\src\Workspaces\Remote\ServiceHub\Shared\RemoteEndPoint.cs:line 234
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at Microsoft.CodeAnalysis.Remote.CodeAnalysisService.<>c__DisplayClass17_0.<<CalculateDiagnosticsAsync>b__0>d.MoveNext() in C:\roslyn\src\Workspaces\Remote\ServiceHub\Services\CodeAnalysisService_Diagnostics.cs:line 48
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Remote.ServiceBase.<RunServiceAsync>d__15.MoveNext() in C:\roslyn\src\Workspaces\Remote\ServiceHub\Shared\ServiceBase.cs:line 110
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.IO.IOException: Pipe is broken.
   at System.IO.Pipes.PipeStream.CheckWriteOperations()
   at System.IO.Pipes.PipeStream.Write(Byte[] buffer, Int32 offset, Int32 count)
   at System.IO.BufferedStream.FlushWrite()
   at System.IO.BufferedStream.Flush()
   at System.IO.BufferedStream.Dispose(Boolean disposing)
   at System.IO.Stream.Close()
   at System.IO.Stream.Dispose()
   at Microsoft.CodeAnalysis.Remote.RemoteEndPoint.<WriteDataToNamedPipeAsync>d__23`1.MoveNext() in C:\roslyn\src\Workspaces\Remote\ServiceHub\Shared\RemoteEndPoint.cs:line 234
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at Microsoft.CodeAnalysis.Remote.CodeAnalysisService.<>c__DisplayClass17_0.<<CalculateDiagnosticsAsync>b__0>d.MoveNext() in C:\roslyn\src\Workspaces\Remote\ServiceHub\Services\CodeAnalysisService_Diagnostics.cs:line 48
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Remote.ServiceBase.<RunServiceAsync>d__15.MoveNext() in C:\roslyn\src\Workspaces\Remote\ServiceHub\Shared\ServiceBase.cs:line 110
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)<---
, CalculateDiagnosticsAsync, 365, Newtonsoft.Json.Linq.JToken[]
```

Seems like disposing the below stream is likely throwing? https://github.com/dotnet/roslyn/blob/d5c6ee52bfc60134c3622f2c763fd558b2e30c67/src/Workspaces/Remote/ServiceHub/Shared/RemoteEndPoint.cs#L250

**Comment by mavasani** [comment]
@tmat

**Comment by mavasani** [comment]
Attaching logs: [servicehub.zip](https://github.com/dotnet/roslyn/files/4898797/servicehub.zip)
The exception traces at 10:45 AM point to this specific issue.

**Comment by tmat** [comment]
The actual issue is with solution snapshot syncrhonization:

```
System.Collections.Generic.KeyNotFoundException: The given key was not present in the dictionary.
   at System.Collections.Concurrent.ConcurrentDictionary`2.get_Item(TKey key)
   at Microsoft.CodeAnalysis.Execution.AssetStorages.<GetRemotableDataAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Execution.RemotableDataServiceFactory.Service.<GetRemotableDataAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Remote.RemoteHostAssetSerialization.<WriteDataAsync>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Remote.RemoteEndPoint.<>c__DisplayClass22_0`1.<<WriteDataToNamedPipeAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Remote.RemoteEndPoint.<WriteDataToNamedPipeAsync>d__23`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.LanguageServices.Remote.ServiceHubRemoteHostClient.<GetAssetsAsync>d__21.MoveNext()
```

**Comment by mavasani** [comment]
Ah, thanks! Let me know if this is a dupe of known issue around solution snapshot being out of sync.

**Comment by CyrusNajmabadi** [comment]
@tmat @sharwell was this fixed by any of your pipe/synchronization fixes?

**Comment by sharwell** [comment]
Yes, will find it

**Comment by sharwell** [comment]
This was fixed by #49371
